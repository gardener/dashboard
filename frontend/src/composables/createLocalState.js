//
// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import { effectScope } from 'vue'

export function createLocalState (stateFactory) {
  let state
  const scope = effectScope(true)

  return (...args) => {
    state = scope.run(() => stateFactory(...args))
    return state
  }
}
