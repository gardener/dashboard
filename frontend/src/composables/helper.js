//
// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import {
  isEmpty,
  isNil,
  isObject,
} from '@/lodash'

export function cleanup (obj) {
  const cleanupObject = obj => {
    const cleanObj = {}
    for (const [key, value] of Object.entries(obj)) {
      const cleanValue = cleanupValue(value)
      if (cleanValue !== null) {
        cleanObj[key] = cleanValue
      }
    }
    return cleanObj
  }

  const cleanupArray = values => {
    const cleanValues = []
    for (const value of values) {
      const cleanValue = cleanupValue(value)
      if (cleanValue !== null) {
        cleanValues.push(cleanValue)
      }
    }
    return cleanValues
  }

  const cleanupValue = value => {
    if (isNil(value)) {
      return null
    }
    if (!isObject(value)) {
      return value
    }
    if (isEmpty(value)) {
      return null
    }
    const obj = Array.isArray(value)
      ? cleanupArray(value)
      : cleanupObject(value)
    if (isEmpty(obj)) {
      return null
    }
    return obj
  }

  return cleanupValue(obj)
}
