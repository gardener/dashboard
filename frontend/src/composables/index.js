//
// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

export { useApi } from './useApi'
export { useAuthn } from './useAuthn'
export { useLogger } from './useLogger'
export { useNav } from './useNav'
export { useTheme } from './useTheme'
export { useToken } from './useToken'
export { useUser } from './useUser'
