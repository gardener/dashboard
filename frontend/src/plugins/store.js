//
// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import pinia from '@/store'

export default {
  install (app) {
    app.use(pinia)
  },
}
