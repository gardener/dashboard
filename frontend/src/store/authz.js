//
// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import {
  defineStore,
  acceptHMRUpdate,
} from 'pinia'
import {
  ref,
  computed,
} from 'vue'

import { useApi } from '@/composables/useApi'

import { canI } from '@/utils'

import { useConfigStore } from './config'
import { useAuthnStore } from './authn'
import { useProjectStore } from './project'

export const useAuthzStore = defineStore('authz', () => {
  const api = useApi()
  const authnStore = useAuthnStore()
  const configStore = useConfigStore()
  const projectStore = useProjectStore()

  const spec = ref(null)
  const status = ref(null)

  const namespace = computed(() => {
    return spec.value?.namespace
  })

  const canCreateTerminals = computed(() => {
    return canI(status.value, 'create', 'dashboard.gardener.cloud', 'terminals')
  })

  const canCreateShoots = computed(() => {
    return canI(status.value, 'create', 'core.gardener.cloud', 'shoots')
  })

  const canPatchShoots = computed(() => {
    return canI(status.value, 'patch', 'core.gardener.cloud', 'shoots')
  })

  const canDeleteShoots = computed(() => {
    return canI(status.value, 'delete', 'core.gardener.cloud', 'shoots')
  })

  const canPatchShootsBinding = computed(() => {
    return canI(status.value, 'patch', 'core.gardener.cloud', 'shoots/binding')
  })

  const canGetSecrets = computed(() => {
    return canI(status.value, 'list', '', 'secrets')
  })

  const canCreateSecrets = computed(() => {
    return canI(status.value, 'create', '', 'secrets')
  })

  const canCreateShootsAdminkubeconfig = computed(() => {
    return canI(status.value, 'create', 'core.gardener.cloud', 'shoots/adminkubeconfig')
  })

  const canCreateShootsViewerkubeconfig = computed(() => {
    return canI(status.value, 'create', 'core.gardener.cloud', 'shoots/viewerkubeconfig')
  })

  const canPatchSecrets = computed(() => {
    return canI(status.value, 'patch', '', 'secrets')
  })

  const canDeleteSecrets = computed(() => {
    return canI(status.value, 'delete', '', 'secrets')
  })

  const canCreateTokenRequest = computed(() => {
    return canI(status.value, 'create', '', 'serviceaccounts/token')
  })

  const canCreateServiceAccounts = computed(() => {
    return canI(status.value, 'create', '', 'serviceaccounts')
  })

  const canPatchServiceAccounts = computed(() => {
    return canI(status.value, 'patch', '', 'serviceaccounts')
  })

  const canDeleteServiceAccounts = computed(() => {
    return canI(status.value, 'delete', '', 'serviceaccounts')
  })

  const canCreateProject = computed(() => {
    return canI(status.value, 'create', 'core.gardener.cloud', 'projects')
  })

  const canPatchProject = computed(() => {
    return canI(status.value, 'patch', 'core.gardener.cloud', 'projects', projectStore.projectName)
  })

  const canManageMembers = computed(() => {
    return canI(status.value, 'manage-members', 'core.gardener.cloud', 'projects', projectStore.projectName)
  })

  const canManageServiceAccountMembers = computed(() => {
    return canPatchProject.value || canManageMembers.value
  })

  const canDeleteProject = computed(() => {
    return canI(status.value, 'delete', 'core.gardener.cloud', 'projects', projectStore.projectName)
  })

  const canGetProjectTerminalShortcuts = computed(() => {
    return canGetSecrets.value
  })

  const canUseProjectTerminalShortcuts = computed(() => {
    return configStore.isProjectTerminalShortcutsEnabled &&
      canGetProjectTerminalShortcuts.value &&
      canCreateTerminals.value
  })

  const hasGardenTerminalAccess = computed(() => {
    return configStore.isTerminalEnabled &&
    canCreateTerminals.value &&
    canPatchServiceAccounts.value &&
    canCreateServiceAccounts.value
  })

  const hasControlPlaneTerminalAccess = computed(() => {
    return configStore.isTerminalEnabled &&
      canCreateTerminals.value &&
      authnStore.isAdmin
  })

  const hasShootTerminalAccess = computed(() => {
    return configStore.isTerminalEnabled &&
      canCreateTerminals.value
  })

  // reuse function not exported
  async function getRules (namespace) {
    const body = { namespace }
    const response = await api.getSubjectRules(body)
    status.value = response.data
  }

  async function fetchRules (namespace) {
    /**
     * The value of `spec.value?.namespace` is:
     * - undefined if no rules have been fetched yet
     * - null if only cluster scoped rules have been fetched
     * - a non-empty string if both cluster scoped rules and the rules for the namespace have been fetched
     */
    if (!namespace) {
      namespace = null
    }
    if (spec.value?.namespace !== namespace) {
      await getRules(namespace)
      this.setNamespace(namespace)
    }
  }

  function refreshRules () {
    return getRules(spec.value?.namespace)
  }

  function setNamespace (namespace) {
    spec.value = { namespace }
  }

  function $reset () {
    spec.value = null
    status.value = null
  }

  return {
    namespace,
    setNamespace,
    canCreateTerminals,
    canCreateShoots,
    canPatchShoots,
    canDeleteShoots,
    canPatchShootsBinding,
    canGetSecrets,
    canCreateSecrets,
    canCreateShootsAdminkubeconfig,
    canCreateShootsViewerkubeconfig,
    canPatchSecrets,
    canDeleteSecrets,
    canCreateTokenRequest,
    canCreateServiceAccounts,
    canPatchServiceAccounts,
    canDeleteServiceAccounts,
    canCreateProject,
    canPatchProject,
    canDeleteProject,
    canManageMembers,
    canManageServiceAccountMembers,
    canGetProjectTerminalShortcuts,
    canUseProjectTerminalShortcuts,
    hasGardenTerminalAccess,
    hasControlPlaneTerminalAccess,
    hasShootTerminalAccess,
    fetchRules,
    refreshRules,
    $reset,
  }
})

if (import.meta.hot) {
  import.meta.hot.accept(acceptHMRUpdate(useAuthzStore, import.meta.hot))
}
