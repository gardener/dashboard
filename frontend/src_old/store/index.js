//
// SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import { listProjectTerminalShortcuts } from '@/utils/api'

// actions
export const actions = {
  fetchCloudProfiles ({ dispatch }) {
    return dispatch('cloudProfiles/getAll')
      .catch(err => {
        dispatch('setError', err)
      })
  },
  async fetchGardenerExtensions ({ dispatch }) {
    try {
      await dispatch('gardenerExtensions/getAll')
    } catch (err) {
      dispatch('setError', err)
    }
  },
  async fetchSeeds ({ dispatch }) {
    try {
      await dispatch('seeds/getAll')
    } catch (err) {
      dispatch('setError', err)
    }
  },
  fetchProjects ({ dispatch }) {
    return dispatch('projects/getAll')
      .catch(err => {
        dispatch('setError', err)
      })
  },
  fetchMembers ({ dispatch, commit }) {
    return dispatch('members/getAll')
      .catch(err => {
        dispatch('setError', err)
      })
  },
  fetchcloudProviderSecrets ({ dispatch, commit }) {
    return dispatch('cloudProviderSecrets/getAll')
      .catch(err => {
        dispatch('setError', err)
      })
  },
  getShootInfo ({ dispatch, commit }, { name, namespace }) {
    return dispatch('shoots/getInfo', { name, namespace })
      .catch(err => {
        dispatch('setError', err)
      })
  },
  setSelectedShoot ({ dispatch }, metadata) {
    return dispatch('shoots/setSelection', metadata)
      .catch(err => {
        dispatch('setError', err)
      })
  },
  async setShootListFilters ({ dispatch, getters }, value) {
    try {
      await dispatch('shoots/setShootListFilters', value)
    } catch (err) {
      dispatch('setError', err)
    }
  },
  async setShootListFilter ({ dispatch, getters }, { filter, value }) {
    try {
      await dispatch('shoots/setShootListFilter', { filter, value })
    } catch (err) {
      dispatch('setError', err)
    }
  },
  createProject ({ dispatch, commit }, data) {
    return dispatch('projects/create', data)
      .then(res => {
        dispatch('setAlert', { message: 'Project created', type: 'success' })
        return res
      })
  },
  patchProject ({ dispatch, commit }, data) {
    return dispatch('projects/patch', data)
  },
  updateProject ({ dispatch, commit }, data) {
    return dispatch('projects/update', data)
      .then(res => {
        dispatch('setAlert', { message: 'Project updated', type: 'success' })
        return res
      })
  },
  deleteProject ({ dispatch, commit }, data) {
    return dispatch('projects/delete', data)
      .then(res => {
        dispatch('setAlert', { message: 'Project deleted', type: 'success' })
        return res
      })
  },
  createCloudProviderSecret ({ dispatch, commit }, data) {
    return dispatch('cloudProviderSecrets/create', data)
      .then(res => {
        dispatch('setAlert', { message: 'Cloud Provider secret created', type: 'success' })
        return res
      })
  },
  updateCloudProviderSecret ({ dispatch, commit }, data) {
    return dispatch('cloudProviderSecrets/update', data)
      .then(res => {
        dispatch('setAlert', { message: 'Cloud Provider secret updated', type: 'success' })
        return res
      })
  },
  deleteCloudProviderSecret ({ dispatch, commit }, data) {
    return dispatch('cloudProviderSecrets/delete', data)
      .then(res => {
        dispatch('setAlert', { message: 'Cloud Provider secret deleted', type: 'success' })
        return res
      })
  },
  createShoot ({ dispatch, commit }, data) {
    return dispatch('shoots/create', data)
      .then(res => {
        dispatch('setAlert', { message: 'Cluster created', type: 'success' })
        return res
      })
  },
  deleteShoot ({ dispatch, commit }, { name, namespace }) {
    return dispatch('shoots/delete', { name, namespace })
      .then(res => {
        dispatch('setAlert', { message: 'Cluster marked for deletion', type: 'success' })
        return res
      })
  },
  subscribeShoots ({ dispatch, commit }) {
    (async () => {
      try {
        await dispatch('shoots/subscribe')
      } catch (err) {
        commit('SET_ALERT', { message: err.message, type: 'error' })
      }
    })()
  },
  unsubscribeShoots ({ dispatch, commit }) {
    (async () => {
      try {
        await dispatch('shoots/unsubscribe')
      } catch (err) {
        commit('SET_ALERT', { message: err.message, type: 'error' })
      }
    })()
  },
  async addMember ({ dispatch, commit }, payload) {
    const result = await dispatch('members/add', payload)
    await dispatch('setAlert', { message: 'Member added', type: 'success' })
    return result
  },
  async updateMember ({ dispatch, commit }, payload) {
    const result = await dispatch('members/update', payload)
    await dispatch('setAlert', { message: 'Member updated', type: 'success' })
    return result
  },
  async deleteMember ({ dispatch, commit }, payload) {
    try {
      const result = await dispatch('members/delete', payload)
      await dispatch('setAlert', { message: 'Member deleted', type: 'success' })
      return result
    } catch (err) {
      await dispatch('setError', { message: `Delete member failed. ${err.message}` })
    }
  },
  async resetServiceAccount ({ dispatch, commit }, payload) {
    try {
      const result = await dispatch('members/resetServiceAccount', payload)
      await dispatch('setAlert', { message: 'Service Account Reset', type: 'success' })
      return result
    } catch (err) {
      await dispatch('setError', { message: `Failed to Reset Service Account ${err.message}` })
    }
  },
  async setNamespace ({ dispatch, commit, state }, namespace) {
    if (namespace && state.namespace !== namespace) {
      commit('SET_NAMESPACE', namespace)
      await dispatch('refreshSubjectRules', namespace)
    }
    return state.namespace
  },
  async ensureProjectTerminalShortcutsLoaded ({ commit, dispatch, state }) {
    const { namespace, projectTerminalShortcuts } = state
    if (!projectTerminalShortcuts || projectTerminalShortcuts.namespace !== namespace) {
      try {
        const { data: items } = await listProjectTerminalShortcuts({ namespace })
        commit('SET_PROJECT_TERMINAL_SHORTCUTS', {
          namespace,
          items,
        })
      } catch (err) {
        // eslint-disable-next-line no-console
        console.warn('Failed to list project terminal shortcuts:', err.message)
      }
    }
  },
}
