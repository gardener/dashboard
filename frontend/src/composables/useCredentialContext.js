//
// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import {
  ref,
  shallowRef,
  computed,
  inject,
  provide,
  watch,
  watchEffect,
} from 'vue'

import { useAuthzStore } from '@/store/authz'

import { cleanup } from '@/composables/helper'

import {
  decodeBase64,
  encodeBase64,
} from '@/utils'

import { useObjectMetadata } from './useObjectMetadata'

import cloneDeep from 'lodash/cloneDeep'
import get from 'lodash/get'
import isEqual from 'lodash/isEqual'
import set from 'lodash/set'
import mapValues from 'lodash/mapValues'

export function createCredentialContextComposable (options = {}) {
  const {
    authzStore = useAuthzStore(),
  } = options

  /* SecretBinding */

  /* SecretBinding initial manifest */
  const initialSecretBindingManifest = shallowRef(null)

  const normalizedInitialSecretBindingManifest = computed(() => {
    const object = cloneDeep(initialSecretBindingManifest.value)
    return normalizeSecretBindingManifest(object)
  })

  /* SecretBinding manifest */
  const secretBindingManifest = ref({})

  function normalizeSecretBindingManifest (value) {
    const object = Object.assign({
      apiVersion: 'core.gardener.cloud/v1beta1',
      kind: 'SecretBinding',
    }, value)
    return cleanup(object)
  }

  const normalizedSecretBindingManifest = computed(() => {
    const object = cloneDeep(secretBindingManifest.value)
    return normalizeSecretBindingManifest(object)
  })

  function setSecretBindingManifest (value) {
    initialSecretBindingManifest.value = value
    secretBindingManifest.value = cloneDeep(initialSecretBindingManifest.value)
  }

  function createSecretBindingManifest () {
    secretBindingManifest.value = {
      metadata: {
        name: '',
        namespace: get(options, ['namespace'], authzStore.namespace),
      },
      provider: {
        type: '',
      },
      secretRef: {
        name: '',
        namespace: get(options, ['namespace'], authzStore.namespace),
      },
    }
    initialSecretBindingManifest.value = cloneDeep(secretBindingManifest.value)
  }

  const isSecretBindingDirty = computed(() => {
    return !isEqual(normalizedSecretBindingManifest.value, normalizedInitialSecretBindingManifest.value)
  })

  /* SecretBinding metadata */
  const secretBindingMetadataComposable = useObjectMetadata(secretBindingManifest)
  const {
    name: secretBindingName,
    namespace: secretBindingNamespace,
  } = secretBindingMetadataComposable

  /* SecretBinding provider */
  const secretBindingProviderType = computed({
    get () {
      return get(secretBindingManifest.value, ['provider', 'type'])
    },
    set (value) {
      set(secretBindingManifest.value, ['provider', 'type'], value || undefined)
    },
  })

  /* SecretBinding secretRef */
  const secretBindingSecretRef = computed({
    get () {
      return get(secretBindingManifest.value, ['secretRef'])
    },
    set (value) {
      set(secretBindingManifest.value, ['secretRef'], value || undefined)
    },
  })

  /* Secret */

  /* Secret initial manifest */
  const initialSecretManifest = shallowRef(null)

  const normalizedInitialSecretManifest = computed(() => {
    const object = cloneDeep(initialSecretManifest.value)
    return normalizeSecretManifest(object)
  })

  /* Secret manifest */
  const secretManifest = ref({})

  function normalizeSecretManifest (value) {
    const object = Object.assign({
      apiVersion: 'v1',
      kind: 'Secret',
      type: 'Opaque',
    }, value)
    return cleanup(object)
  }

  const normalizedSecretManifest = computed(() => {
    const object = cloneDeep(secretManifest.value)
    return normalizeSecretManifest(object)
  })

  function setSecretManifest (value) {
    initialSecretManifest.value = value
    secretManifest.value = cloneDeep(initialSecretManifest.value)
  }

  function createSecretManifest () {
    secretManifest.value = {
      metadata: {
        name: '',
        namespace: get(options, ['namespace'], authzStore.namespace),
      },
      data: {},
    }
    initialSecretManifest.value = cloneDeep(secretManifest.value)
  }

  const isSecretDirty = computed(() => {
    return !isEqual(normalizedSecretManifest.value, normalizedInitialSecretManifest.value)
  })

  /* Secret metadata */
  const secretMetadataComposable = useObjectMetadata(secretManifest)
  const {
    name: secretName,
    namespace: secretNamespace,
  } = secretMetadataComposable

  /* Secret data */
  const secretData = computed({
    get () {
      return get(secretManifest.value, ['data'])
    },
    set (value) {
      set(secretManifest.value, ['data'], value || undefined)
    },
  })

  const secretStringData = computed({
    get () {
      return mapValues(secretData.value, decodeBase64)
    },
    set (value) {
      secretData.value = value ? mapValues(value, encodeBase64) : undefined
    },
  })

  function secretStringDataRefs (keyMapping) {
    const refs = {}

    for (const [, variableName] of Object.entries(keyMapping)) {
      set(refs, [variableName], ref(''))
    }

    watch(
      secretStringData,
      newValue => {
        if (newValue) {
          for (const [dataKey, variableName] of Object.entries(keyMapping)) {
            set(refs, [variableName, 'value'], get(newValue, [dataKey], ''))
          }
        }
      },
      { immediate: true, deep: true },
    )

    watchEffect(() => {
      secretStringData.value = Object.fromEntries(
        Object.entries(keyMapping).map(([dataKey, variableName]) => [
          dataKey,
          get(refs, [variableName, 'value']),
        ]),
      )
    })

    return refs
  }

  return {
    /* SecretBinding */
    secretBindingManifest: normalizedSecretBindingManifest,
    setSecretBindingManifest,
    createSecretBindingManifest,
    isSecretBindingDirty,
    secretBindingName,
    secretBindingNamespace,
    secretBindingProviderType,
    secretBindingSecretRef,

    /* Secret */
    secretManifest: normalizedSecretManifest,
    setSecretManifest,
    createSecretManifest,
    isSecretDirty,
    secretName,
    secretStringData,
    secretStringDataRefs,
    secretNamespace,
    secretData,
  }
}

export function useCredentialContext () {
  return inject('credential-context', null)
}

export function useProvideCredentialContext (options) {
  const composable = createCredentialContextComposable(options)
  provide('credential-context', composable)
  return composable
}
