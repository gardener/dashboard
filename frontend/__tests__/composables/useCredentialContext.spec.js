//
// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import {
  reactive,
  nextTick,
} from 'vue'
import {
  setActivePinia,
  createPinia,
} from 'pinia'

import { useAuthzStore } from '@/store/authz'

import { createCredentialContextComposable } from '@/composables/useCredentialContext'

describe('composables', () => {
  describe('useCredentialContext', () => {
    const testNamespace = 'garden-foo'
    let credentialContext

    beforeEach(() => {
      setActivePinia(createPinia())
      const authzStore = useAuthzStore()
      authzStore.setNamespace(testNamespace)

      const composable = createCredentialContextComposable()
      credentialContext = reactive(composable)
    })

    it('should set SecretBinding manifest', () => {
      const manifest = { metadata: { name: 'foo', namespace: testNamespace }, provider: { type: 'aws' } }
      credentialContext.setSecretBindingManifest(manifest)

      expect(credentialContext.secretBindingName).toBe('foo')
      expect(credentialContext.secretBindingNamespace).toBe(testNamespace)
      expect(credentialContext.secretBindingProviderType).toBe('aws')
      expect(credentialContext.secretBindingManifest).toMatchSnapshot()
    })

    it('should create a new SecretBinding manifest with defaults', () => {
      credentialContext.createSecretBindingManifest()
      expect(credentialContext.secretBindingName).toBe('')
      expect(credentialContext.secretBindingNamespace).toBe(testNamespace)
      expect(credentialContext.secretBindingProviderType).toBe('')
      expect(credentialContext.secretBindingManifest).toMatchSnapshot()
    })

    it('should detect dirty state in SecretBinding', () => {
      credentialContext.createSecretBindingManifest()
      expect(credentialContext.isSecretBindingDirty).toBe(false)
      credentialContext.secretBindingProviderType = 'gcp'
      expect(credentialContext.isSecretBindingDirty).toBe(true)
    })

    it('should set and create Secret manifest', () => {
      const secretManifest = {
        metadata: { name: 'my-secret', namespace: testNamespace },
        data: { foo: 'YmFy' },
      }
      credentialContext.setSecretManifest(secretManifest)
      expect(credentialContext.secretName).toBe('my-secret')
      expect(credentialContext.secretData).toEqual({ foo: 'YmFy' })

      credentialContext.createSecretManifest()
      expect(credentialContext.secretName).toBe('')
      expect(credentialContext.secretData).toEqual({})
    })

    it('should detect dirty state in Secret', () => {
      credentialContext.createSecretManifest()
      expect(credentialContext.isSecretDirty).toBe(false)
      credentialContext.secretName = 'changed-secret'
      expect(credentialContext.isSecretDirty).toBe(true)
    })

    it('should update and encode/decode secretStringData', () => {
      credentialContext.createSecretManifest()
      credentialContext.secretStringData = { password: 'secret' }
      expect(credentialContext.secretData).toEqual({ password: 'c2VjcmV0' })
      expect(credentialContext.secretStringData).toEqual({ password: 'secret' })
    })

    it('should update secret data via secretStringDataRefs', async () => {
      credentialContext.setSecretManifest({ metadata: { name: 'my-secret', namespace: testNamespace }, data: { password: 'aW5pdGlhbA==' } })

      const keyMapping = { password: 'pwdVar', token: 'tokenVar' }
      const refs = credentialContext.secretStringDataRefs(keyMapping)

      await nextTick()
      expect(refs.pwdVar.value).toBe('initial')

      refs.pwdVar.value = 'mypassword'
      refs.tokenVar.value = 'mytoken'

      await nextTick()
      expect(credentialContext.secretStringData).toEqual({
        password: 'mypassword',
        token: 'mytoken',
      })
      expect(credentialContext.secretData).toEqual({
        password: 'bXlwYXNzd29yZA==',
        token: 'bXl0b2tlbg==',
      })
    })

    it('should update SecretBinding provider type and secretRef', () => {
      credentialContext.createSecretBindingManifest()
      credentialContext.secretBindingProviderType = 'azure'
      expect(credentialContext.secretBindingProviderType).toBe('azure')

      credentialContext.secretBindingSecretRef = { name: 'my-secret-ref', namespace: 'other-namespace' }
      expect(credentialContext.secretBindingSecretRef).toEqual({ name: 'my-secret-ref', namespace: 'other-namespace' })
    })
  })
})
