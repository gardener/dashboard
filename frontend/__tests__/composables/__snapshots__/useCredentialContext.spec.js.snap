// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`composables > useCredentialContext > should create a new SecretBinding manifest with defaults 1`] = `
{
  "apiVersion": "core.gardener.cloud/v1beta1",
  "kind": "SecretBinding",
  "metadata": {
    "name": "",
    "namespace": "garden-foo",
  },
  "provider": {
    "type": "",
  },
  "secretRef": {
    "name": "",
    "namespace": "garden-foo",
  },
}
`;

exports[`composables > useCredentialContext > should set SecretBinding manifest 1`] = `
{
  "apiVersion": "core.gardener.cloud/v1beta1",
  "kind": "SecretBinding",
  "metadata": {
    "name": "foo",
    "namespace": "garden-foo",
  },
  "provider": {
    "type": "aws",
  },
}
`;
